File: src/main/java/me/piperswe/demogui/LabelComponent.java
package me.piperswe.demogui;

import javafx.scene.control.Label;
import lombok.NonNull;
import me.piperswe.utils.gui.ReactiveComponent;
import me.piperswe.utils.gui.StateContainer;
import org.jetbrains.annotations.NotNull;

public class LabelComponent extends ReactiveComponent<State, Integer> {
    private final Label label = new Label();

    public LabelComponent(@NonNull StateContainer<State> stateContainer) {
        super(stateContainer);
    }

    @Override
    public Label render() {
        return label;
    }

    @NotNull
    @Override
    public Integer select(@NotNull State state) {
        return state.counter();
    }

    @Override
    public void handleStateChange(Integer newState, Integer oldState) {
        String javaVersion = System.getProperty("java.version");
        String javafxVersion = System.getProperty("javafx.version");
        label.setText(String.format(
                "Hello, JavaFX %s (running on Java %s)! %d click(s) so far.",
                javafxVersion, javaVersion, newState
        ));
    }
}

File: src/main/kotlin/me/piperswe/demogui/kotlin/LabelComponent.kt
package me.piperswe.demogui.kotlin

import javafx.scene.control.Label
import me.piperswe.utils.gui.createReactiveComponentFactory

val labelFactory = createReactiveComponentFactory<State, Int, Label>(
    { Label() },
    { it.counter },
    { counter, _, label ->
        val javaVersion = System.getProperty("java.version")
        val javafxVersion = System.getProperty("javafx.version")
        label.text =
            "Hello, JavaFX $javafxVersion (running on Java $javaVersion, Kotlin ${KotlinVersion.CURRENT})! $counter click(s) so far."
    }
)